keps/1397-evictions-in-background/README.md

```diff
@@ -0,0 +1,1122 @@
+existing implementation, the annotation based eviction can be seen
+as an v1alpha1 implemenation of this proposal.
+Integration with the evacuation API as an v1alphaN or beta implementation.
+
+### Workflow Description
+
+This section outlines the workflow for enabling the proposed evictions
+in background that prevent workloads from being evicted right away
+without following additional eviction policies.
+
+#### Actors
+
+- Cluster Administrator: Responsible for configuring descheduling
+  policies and overseeing cluster operations.
+- User: Individuals or services running workloads within the cluster.
+- Descheduler: Component responsible for requesting pod evictions
+
+#### Workflow Steps
+
+1. Policy configuration
+   1. The cluster administrator configures the descheduler to enable the new functionality.
+2. A user deploys a workload through various nodes
+   1. A scheduler distributes the pods based on configured scheduling policies
+3. Eviction nomination
+   1. At the beginning of each descheduling cycle the descheduler increases
+      all the internal counters to take into account all pods that are subjects
+      to background eviction. Either by checking `descheduler.alpha.kubernetes.io/eviction-in-progress` annotation,
+      list of evacuation requests or confronting the internal caches.
+   2. Each descheduling plugin nominates a set of pods to be evicted.
+   3. All the nominated pods are sorted based on the eviction-in-progress first priority.
+4. Pod eviction: the descheduler starts evicting nominated pods until a limit is reached
+   1. If an annotation based eviction is enabled:
+      1. If a pod is not annotated with `descheduler.alpha.kubernetes.io/request-evict-only` evict
+         the pod normally without additional handling.
+      2. If a pod is annotated with `descheduler.alpha.kubernetes.io/request-evict-only` evict the pod.
+         1. If the eviction fails intepret the error as a request for eviction in background.
+         2. Error code, resp. response text is checked to distinguish between
+            a genuine error and a confirmation of an eviction in background.
+   2. If the evacuation API is utilized create an evacution request
+
+### Notes/Constraints/Caveats (Optional)
+
+- The upstream eviction API does not currently support evictions in background
+  implemented by external components. For that, there's no community
+  dedicated error code nor a response text for this type of eviction.
+- The eviction in background error code will be temporary mapped to 429 (TooManyRequests).
+- The response text will be searched for strings
+  containing "Eviction triggered" prefix or similar.
+- The descheduler will still treat all the pods with eviction in background as cattle.
+  It will only prioritize these pods to be processed sooner.
+  There may still be unnecessary/sub-optimal evictions.
+
+### Risks and Mitigations
+
+- If an external component responsible for eviction in background
```

Comment:
does this mean that we will retry the eviction request? E.g. for the case that the eviction information was lost by the evacuator? How is this and the TTL going to look in beta?

Reply:
If the descheduler is the only evacuation requester, the evacuation request gets deleted after TTL. So other pods have a chance to be evacuated. Depending on how pods are proceed in the next descheduling cycle the same evacuation request can be re-created.